### PR TITLE
Add Sparse VideoGen attention for Wan on TPUs

### PR DESCRIPTION
**Pull request by Ravi-VK**
Video diffusion processes long sequences of video tokens across many denoising steps, making self-attention expensive. Sparse attention reduces this cost by skipping interactions that contribute little to the output.

This PR introduces an opt-in sparse attention implementation for MaxDiffusion: [Sparse VideoGen (SVG)](https://arxiv.org/abs/2502.01776) for Wan on TPUs. SVG samples a few queries to choose a spatial or temporal pattern for each head, then computes only the selected mask. Density and active steps/layers are configurable, and SVG is disabled by default.

See [SVG.md](https://github.com/Ravi-VK/maxdiffusion-svg/blob/svg-head-local-pr-clean/docs/svg.md) for an illustrated explanation, configuration examples, and profiling instructions.

The implementation rounds sparse boundaries to hardware tiles, approximately preserving the attention-pair budget. Selected interior tiles share one mask-free kernel; only tiles touching sequence padding need validity masks.

The feature is organized into six commits:

1. **[Kernel primitives](https://github.com/Ravi-VK/maxdiffusion-svg/commit/4907c99f9916175dde26db1a46b194a3e23773e8):** compute attention over selected ranges of tiles.
2. **[Tile rounding and dispatch](https://github.com/Ravi-VK/maxdiffusion-svg/commit/3b9f325508b2ae60546fabf30eeb00dea3393b7f):** select boundary tiles and combine main and padding-tail results.
3. **[Routing and layouts](https://github.com/Ravi-VK/maxdiffusion-svg/commit/6db9481632c7a8072602f41197d4a34bbd778587):** choose each head’s pattern and arrange tokens for efficient execution.
4. **[Wan integration](https://github.com/Ravi-VK/maxdiffusion-svg/commit/719f6ffe9abea85f597c70098963f5e19b8fa1fb):** connect SVG to inference with configuration controls and unsupported-mode checks.
5. **[Tests](https://github.com/Ravi-VK/maxdiffusion-svg/commit/ebc9ae72bb464419c4685ed32a611643a97081fd):** cover routing, layouts, configuration, sharding, and kernel correctness.
6. **[Documentation](https://github.com/Ravi-VK/maxdiffusion-svg/commit/e38cbc43709a2fc13f860e370cb1193d692d9473):** explain the method, usage, profiling, and results with illustrations.

A [review follow-up](https://github.com/Ravi-VK/maxdiffusion-svg/commit/7ea23be0c0b04627ac3b014621571ac0f87b5060) adds configuration and shape guards, a static inactive-step shortcut, and regression tests.

Wan2.2 720p results on TPU v6e-8, using 81 frames and 40 denoising steps:

| Policy | Estimated transformer FLOPs saved | Denoising speedup | PSNR (dB) |
|---|---:|---:|---:|
| Conservative | ≈27.2% | **1.13×** | 26.47 |
| Moderate | ≈32.2% | **1.20×** | 26.14 |
| Aggressive | ≈37.3% | **1.28×** | 24.80 |

Timings are medians of three warm runs against same-node optimized fixed-M dense controls. PSNR uses FFmpeg’s aggregate YUV metric against dense outputs.

In an earlier evaluation using the moderate SVG policy, Wan2.2 at 720p retained **98.1% of the dense baseline’s mean VBench dimension score** across a 31-prompt, 16-dimension screening subset.

SVG supports inference through the custom Ulysses/ring backends; training, Animate, external masks, periodic support, chunked Ulysses, CFG cache, and MagCache are unsupported.

### CODE REVIEW

File: src/maxdiffusion/kernels/custom_svg_balanced_rounding_attention.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+      items,
+      key=lambda x: (-x.alpha, -x.exact_pairs, x.qi, x.kj),
+  )
+  best_k, best_err, running = 0, abs(float(target)), 0
+  for k, tile in enumerate(ordered, 1):
+    running += tile.real_pairs
+    err = abs(float(running) - float(target))
+    if err < best_err:
+      best_k, best_err = k, err
+  return {(x.qi, x.kj) for x in ordered[:best_k]}
```
Comment (Perseus14) [inline src/maxdiffusion/kernels/custom_svg_balanced_rounding_attention.py:151]: `_closest_prefix()` minimizes the global pair-budget error without ensuring each query tile retains any support. It can select zero boundary tiles even when there are no full tiles.

When `Q = K = 0` and every value equal to one, the kernel returns all zeros. Attention over any nonempty set of these values must return ones.

This also occurs for individual query rows on larger grids at low densities. Preserving the global budget does not guarantee valid attention locally.

File: src/maxdiffusion/pipelines/wan/wan_pipeline_2_2.py [hunk 2]
```diff
@@ -193,6 +193,11 @@ def __call__(
           "SenCache requires classifier-free guidance to be enabled for both transformer phases."
       )
 
+    low_noise_config = getattr(self.low_noise_transformer, "config", None)
+    if getattr(self, "use_svg_attention", False) or getattr(low_noise_config, "use_svg_attention", False):
+      if use_cfg_cache or use_magcache:
+        raise ValueError("SVG sparse attention cannot be combined with CFG cache or MagCache.")
+
```
Comment (Perseus14) [inline src/maxdiffusion/pipelines/wan/wan_pipeline_2_2.py:200]: `use_svg_attention` is in `attention_config` and not `transformer.config` as per https://github.com/AI-Hypercomputer/maxdiffusion/blob/e38cbc43709a2fc13f860e370cb1193d692d9473/src/maxdiffusion/pipelines/wan/wan_pipeline.py#L354-L363

File: src/maxdiffusion/models/wan/transformers/svg_attention.py [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+  `scale` is the ordinary attention logit scale. The profiler intentionally
+  uses natural-exp softmax. The production base-2 kernel multiplies Q by
+  log2(e) before exp2, which is mathematically equivalent to this natural-exp
+  formulation, so no additional LOG2E factor belongs in the profiler.
+  """
+  with jax.named_scope("svg_route_profile"):
+    sequence_length = query.shape[2]
+    sample_pool_size = min(max(int(sample_max_row), 1), sequence_length)
+    sample_count = min(max(int(query_count), 1), sequence_length)
```
Comment (Perseus14) [inline src/maxdiffusion/models/wan/transformers/svg_attention.py:112]: This might be more accurate `sample_count = min(max(int(query_count), 1), sample_pool_size)`

**Review by Perseus14 — COMMENTED** [review]
Added some minor comments. PTAL

File: src/maxdiffusion/models/wan/transformers/svg_attention.py [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+  dense_layer_count = math.ceil(dense_layer_fraction * num_layers)
+  if dense_layer_count > 0 and layer_index is not None:
+    is_active = jnp.logical_and(is_active, jnp.asarray(layer_index) >= dense_layer_count)
+  if dense_timestep_fraction > 0.0 and timestep is not None:
+    first_sparse_timestep = (1.0 - dense_timestep_fraction) * num_train_timesteps
+    is_active = jnp.logical_and(is_active, jnp.max(jnp.asarray(timestep)) < first_sparse_timestep)
+  return is_active
+
+
+def compute_band_width(sequence_length_or_grid: int | Tuple[int, int, int], density: float) -> int:
```
Comment (jitendra-jalwaniya) [inline src/maxdiffusion/models/wan/transformers/svg_attention.py]: This method seems unused. I think the code overall is only using `svg_execution_band_width()`

File: src/maxdiffusion/models/attention_flax.py [hunk 5]
```diff
@@ -402,6 +405,7 @@ def _extract_custom_block_sizes(flash_block_sizes):
   # to 1 (the custom-kernel default) to keep the `heads_per_tile > 1` guards safe.
   if heads_per_tile is None:
     heads_per_tile = 1
+  bkv_compute_in = min(bkv_compute, bkv_compute_in)
```
[the remark this answers is not in the document]
Comment (Ravi-VK) [reply src/maxdiffusion/models/attention_flax.py:408]: `bkv_compute_in` is initialized to 1024, and both configuration paths use `value or bkv_compute_in`, so missing or explicit `None` values retain the numeric default. It therefore cannot be `None` when we reach `min()`.

File: src/maxdiffusion/kernels/custom_svg_balanced_rounding_attention.py [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+    with jax.named_scope("svg_tail_cleanup"):
+      ot, lt = tail(q, k, v)
+    with jax.named_scope("svg_lse_merge"):
+      m = jnp.maximum(lm, lt)
+      exp = jnp.exp2 if use_base2_exp else jnp.exp
+      wm, wt = exp(lm - m), exp(lt - m)
+      den = wm + wt
+      return (om.astype(jnp.float32) * (wm / den)[:, None, :] + ot.astype(jnp.float32) * (wt / den)[:, None, :]).astype(
+          q.dtype
+      )
```
[the remark this answers is not in the document]
Comment (Ravi-VK) [reply src/maxdiffusion/kernels/custom_svg_balanced_rounding_attention.py:360]: Empty partials use a large finite negative value, not -inf, and return zero output. When both are empty, the merge returns zero without NaNs. The separate issue of queries losing all support still needs a fix.

File: src/maxdiffusion/kernels/custom_svg_static_range_attention.py [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+  def attention(q, k, v):
+    of, lf = full(q, k, v)
+    ob, lb = boundary(q, k, v)
+    m = jnp.maximum(lf, lb)
+    exp = jnp.exp2 if use_base2_exp else jnp.exp
+    wf, wb = exp(lf - m), exp(lb - m)
+    den = wf + wb
+    return (of.astype(jnp.float32) * (wf / den)[:, None, :] + ob.astype(jnp.float32) * (wb / den)[:, None, :]).astype(
+        q.dtype
+    )
```
[the remark this answers is not in the document]
Comment (Ravi-VK) [reply src/maxdiffusion/kernels/custom_svg_static_range_attention.py:371]: Same here - empty partials use a large finite negative value, not -inf, and return zero output. When both are empty, the merge returns zero without NaNs.

File: src/maxdiffusion/kernels/custom_svg_balanced_rounding_partial.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+      bq=bq,
+      bkv=bkv,
+  )
+  scalars = tuple(jnp.asarray(x) for x in (table_np, active_np, q_valid, k_valid))
+  qtiles, width = table_np.shape
+
+  def partial(q, k, v):
+    heads, _, dq = q.shape
+    dv = v.shape[-1]
+
```
[the remark this answers is not in the document]
Comment (Ravi-VK) [reply src/maxdiffusion/kernels/custom_svg_balanced_rounding_partial.py:199]: Fixed in 7ea23be. Added the divisibility check and a test that unsupported value dimensions raise an error.

File: src/maxdiffusion/models/wan/transformers/svg_attention.py [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+    num_train_timesteps: int = 1000,
+    num_layers: int = 40,
+) -> bool | jax.Array:
+  """Evaluate whether SVG attention is active for the current step and layer."""
+  explicit_schedule = end_step >= 0
+  if explicit_schedule:
+    if step_index is None or layer_index is None:
+      raise ValueError("Explicit SVG schedules require both step_index and layer_index.")
+    if isinstance(layer_index, Integral) and not start_layer <= layer_index < end_layer:
+      return False
```
[the remark this answers is not in the document]
Comment (Ravi-VK) [reply src/maxdiffusion/models/wan/transformers/svg_attention.py]: Fixed in 7ea23be. The function now returns early when the step is a known inactive Python integer. Added a test with a fixed step and a traced layer index.

File: src/maxdiffusion/pipelines/wan/wan_pipeline_2_2.py [hunk 10]
```diff
@@ -193,6 +193,10 @@ def __call__(
           "SenCache requires classifier-free guidance to be enabled for both transformer phases."
       )
 
+    if getattr(self, "use_svg_attention", False) or getattr(self.low_noise_transformer.config, "use_svg_attention", False):
```
[the remark this answers is not in the document]
Comment (Ravi-VK) [reply src/maxdiffusion/pipelines/wan/wan_pipeline_2_2.py]: Fixed in 7ea23be. This now handles a transformer that is None or has no config attribute. Added a test for the unloaded-transformer case.

File: src/maxdiffusion/pipelines/wan/wan_pipeline.py [hunk 11]
```diff
@@ -342,11 +341,43 @@ def create_model(rngs: nnx.Rngs, wan_config: dict):
   wan_config["mask_padding_tokens"] = config.mask_padding_tokens
   wan_config["scan_layers"] = config.scan_layers
   wan_config["enable_jax_named_scopes"] = config.enable_jax_named_scopes
+  high_density = float(getattr(config, "svg_high_noise_density", -1.0))
+  low_density = float(getattr(config, "svg_low_noise_density", -1.0))
```
[the remark this answers is not in the document]
Comment (Ravi-VK) [reply src/maxdiffusion/pipelines/wan/wan_pipeline.py]: Fixed in 7ea23be. Missing or None expert densities now fall back to svg_spatial_density. Added tests for both experts.

File: src/maxdiffusion/kernels/custom_svg_balanced_rounding_attention.py (hunk 1 above)
Comment (Ravi-VK) [reply src/maxdiffusion/kernels/custom_svg_balanced_rounding_attention.py:153]: Fixed in e5298b8. If rounding leaves a query tile without any keys, we now retain one boundary tile for that row and include its cost in the budget report. Added coverage checks and a TPU regression test for the constant-value example.

File: src/maxdiffusion/pipelines/wan/wan_pipeline_2_2.py (hunk 2 above)
Comment (Ravi-VK) [reply src/maxdiffusion/pipelines/wan/wan_pipeline_2_2.py]: Fixed in e5298b8. The low-noise check now reads use_svg_attention from attention_config. Added a regression test for low-noise SVG with incompatible caches.

File: src/maxdiffusion/models/wan/transformers/svg_attention.py (hunk 3 above)
Comment (Ravi-VK) [reply src/maxdiffusion/models/wan/transformers/svg_attention.py]: Fixed in e5298b8. sample_count is now capped by sample_pool_size.

File: src/maxdiffusion/models/wan/transformers/svg_attention.py (hunk 4 above)
Comment (Ravi-VK) [reply src/maxdiffusion/models/wan/transformers/svg_attention.py]: Removed compute_band_width() in e5298b8. The implementation uses svg_execution_band_width().

**Review by Perseus14 — DISMISSED** [review]

**Review by jitendra-jalwaniya — DISMISSED** [review]

**Comment by Perseus14** [comment]
@Ravi-VK Please squash the commits

File: docs/svg.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+| Moderate | ≈32.2% | 153.37 s | 127.94 s | **1.20×** | **26.14** |
+| Aggressive | ≈37.3% | 153.50 s | 119.86 s | **1.28×** | **24.80** |
+
+*PSNR is measured against dense outputs using FFmpeg's aggregate YUV metric.*
+
+In an earlier evaluation using the moderate SVG policy, Wan2.2 at 720p retained **98.1% of the dense baseline’s mean VBench dimension score** across a 31-prompt, 16-dimension screening subset.
+
+## Qualitative example
+
+![Three rows of giraffe video frames, with dense attention on the left and SVG on the right.](images/svg/dense-vs-svg.png)
```
Comment (mbohlool) [inline docs/svg.md:124]: `docs/images/svg/dense-vs-svg.png` is currently committed at **7.86 MB** (7,861,852 bytes).

While `.pre-commit-config.yaml` enforces a 5 MB limit on newly added files (`check-added-large-files`), updating an existing path bypassed this check. Committing a 7.86 MB image directly into the git tree permanently bloats git clone size and packfile history for everyone.

Could you please optimize/compress this image to bring it under 1 MB?

**Review by mbohlool — CHANGES_REQUESTED** [review]

**Review by jitendra-jalwaniya — APPROVED** [review]

File: docs/svg.md (hunk 12 above)
Comment (Ravi-VK) [reply docs/svg.md:124]: Done - compressed to <1MB and squashed PR into a single commit: [db8c0cd](https://github.com/AI-Hypercomputer/maxdiffusion/commit/db8c0cd45942cbbca9a23865a7ac1e356c62b12e)

**Review by mbohlool — APPROVED** [review]
